src/group_theory/double_coset.lean

```diff
@@ -0,0 +1,226 @@
+This file defines double cosets for two subgroups `H K` of a group `G` and the quotient of `G` by
+the double coset relation, i.e. `H \ G / K`. We also prove that `G` can be writen as a disjoint
+union of the double cosets and that if one of `H` or `K` is the trivial group (i.e. `⊥` ) then
+this is the usual left or right quotient of a group by a subgroup.
+
+## Main definitions
+
+* `rel`: The double coset relation defined by two subgroups `H K` of `G`.
+* `double_coset.quotient`: The quotient of `G` by the double coset relation, i.e, ``H \ G / K`.
+-/
+
+
+variables {G : Type*} [group G] {α : Type*} [has_mul α] (J: subgroup G) (g : G)
+
+namespace doset
+
+open_locale pointwise
+
+/--The double_coset as an element of `set α` corresponding to `s a t` -/
+def _root_.doset (a : α) (s t : set α) : set α := s * {a} * t
+
+lemma mem_doset {s t : set α} {a b : α} : b ∈ doset a s t ↔ ∃ (x ∈ s) (y ∈ t), b = x * a * y :=
+⟨λ ⟨_, y, ⟨x, _, hx, rfl, rfl⟩, hy, h⟩, ⟨x, hx, y, hy, h.symm⟩,
+  λ ⟨x, hx, y, hy, h⟩, ⟨x * a, y, ⟨x, a, hx, rfl, rfl⟩, hy, h.symm⟩⟩
+
+lemma mem_doset_self (H K : subgroup G) (a : G) : a ∈ doset a H K :=
+mem_doset.mpr ⟨1, H.one_mem, 1, K.one_mem, (one_mul a).symm.trans (mul_one (1 * a)).symm⟩
+
+lemma doset_eq_of_mem {H K : subgroup G} {a b : G} (hb : b ∈ doset a H K) :
+  doset b H K = doset a H K :=
+begin
+  obtain ⟨_, k, ⟨h, a, hh, (rfl : _ = _), rfl⟩, hk, rfl⟩ := hb,
+  rw [doset, doset, ←set.singleton_mul_singleton, ←set.singleton_mul_singleton, mul_assoc,
+    mul_assoc, subgroup.singleton_mul_subgroup hk, ←mul_assoc, ←mul_assoc,
+    subgroup.subgroup_mul_singleton hh],
+end
+
+lemma mem_doset_of_not_disjoint {H K : subgroup G} {a b : G}
+  (h : ¬ disjoint (doset a H K) (doset b H K)) : b ∈ doset a H K :=
+begin
+  rw set.not_disjoint_iff at h,
+  simp only [mem_doset] at *,
+  obtain ⟨x, ⟨l, hl, r, hr, hrx⟩, y, hy, ⟨r', hr', rfl⟩⟩ := h,
+  refine ⟨y⁻¹ * l, H.mul_mem (H.inv_mem hy) (hl), r * r'⁻¹, K.mul_mem hr (K.inv_mem hr'), _⟩,
+  rwa [mul_assoc, mul_assoc, eq_inv_mul_iff_mul_eq, ←mul_assoc, ←mul_assoc, eq_mul_inv_iff_mul_eq],
+end
+
+lemma eq_of_not_disjoint {H K : subgroup G} {a b : G} (h: ¬ disjoint (doset a H K) (doset b H K)) :
+  doset a H K = doset b H K :=
+begin
+  rw disjoint.comm at h,
+  have ha : a ∈ doset b H K := mem_doset_of_not_disjoint h,
+  apply doset_eq_of_mem ha,
+end
+
+/--The relation given by saying two elements are related if their double cosets are equal -/
+def rel (H K : set G) : G → G → Prop :=
+λ x y, doset x H K = doset y H K
+
+lemma rel_reflexive {H K : set G} : reflexive (rel H K) :=
+λ x, rfl
+
+lemma rel_symmetric {H K : set G} : symmetric (rel H K) :=
+λ x y, eq.symm
+
+lemma rel_transitive {H K : set G} : transitive (rel H K) :=
+λ x y z, eq.trans
+
+lemma rel_equivalence {H K : set G} : equivalence (rel H K) :=
+⟨rel_reflexive, rel_symmetric, rel_transitive⟩
+
+/-- The setoid defined by the double_coset relation -/
+def setoid (H K : set G) : setoid G :=
+⟨rel H K, rel_equivalence⟩
```

Comment:
You can get here a lot faster with
```suggestion
/-- The setoid defined by the double_coset relation -/
def setoid (H K : set G) : setoid G :=
setoid.ker (λ x, doset x H K)
```

Reply:
On nice. I'll keep the definition of rel as this is needed elsewhere.

Reply:
What do you need `rel` for?

Reply:
Ah I sure I see what you mean. I've removed it now :)